In [63]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

아예리네이미하는게 간편할수도
sepal length 0  
sepal width 1  
petal length 2  
petal width 3  
class 4  
df = df.rename({'sepal length (cm)':'SL', 'sepal width (cm)':'SW', 'petal length (cm)':'PL', 'petal width (cm)':'PW'}, axis='columns')  
df.head()  

범주형자료가 있는데 회귀해야한다면,, 원핫인코딩, 레이블인코딩필요없이 맵핑도 가능  
df'target' = dataset.target  
df'target' = df'target'.map({0:"setosa", 1:"versicolor", 2:"virginica"})  


In [64]:
from sklearn.datasets import load_iris
iris = load_iris()
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [65]:
df['class'] = iris.target
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   class  
0      0  
1      0  
2      0  
3      0  
4      0

In [66]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

## petal length의 weight, bias

In [69]:
X = df.iloc[:,[0,1,3,4]].values
y = df.iloc[:,2].values     #petal length

In [70]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=2021, shuffle=True
)

In [100]:
mul_lr = LinearRegression()
mul_lr.fit(X_train, y_train)
pred = mul_lr.predict(X_test)
mse = mean_squared_error(y_test, pred)
r2 = r2_score(y_test, pred) 
print(f'MSE: {mse:.2f}')
print(f'RMSE: {np.sqrt(mse):.2f}')
print(f'R_squared: {r2:.4f}')

다중 선형 회귀, R2 : 0.5667
다중 선형 회귀, 계수(w) : [ 0.61339502 -0.59499152  0.69803975 -0.12580651], 절편(b) : 0.9977
MSE: 0.05
RMSE: 0.22
R_squared: 0.5667


### 회귀식
-  0.6822\*sepal_length - 0.5192\*sepal_width + 1.1166\*petal_width + 0.3750\*class -0.3670

### 교차검증 - cross_val_score

In [75]:
from sklearn.model_selection import cross_val_score
lr = LinearRegression()
# 5폴드5Fold세트로 MSE를 구함
neg_mean_scores = cross_val_score(lr, X, y, scoring='neg_mean_squared_error', cv=5) #neg_mean_squared_error 때문에 값이 음수로 나온다..
rmse_scores = np.sqrt(-neg_mean_scores)
avg_rmse = np.average(rmse_scores)

print('개별 MSE:', np.round(neg_mean_scores,2))
print('개별 RMSE:', np.round(rmse_scores,2))
print(f'평균 RMSE:, {avg_rmse:.4f}')

개별 MSE: [-0.06 -0.08 -0.14 -0.1  -0.21]
개별 RMSE: [0.25 0.28 0.38 0.32 0.46]
평균 RMSE:, 0.3387


### GridSearch

In [76]:
from sklearn.model_selection import GridSearchCV
lr = LinearRegression()

In [77]:
lr.get_params()

{'copy_X': True, 'fit_intercept': True, 'n_jobs': None, 'normalize': False}

In [78]:
parameters = {'copy_X':[True, False], 'fit_intercept':[True,False], 'normalize':[True,False]}
grid = GridSearchCV(lr,parameters, cv=None)
grid.fit(X_train, y_train)
print("r2 / variance : ", grid.best_score_)
print("Residual sum of squares: %.2f"
              % np.mean((grid.predict(X_test) - y_test) ** 2))
print(f'최고 평균 정확도: {grid.best_score_:.4f}')
print('최적 파라미터:', grid.best_params_)

r2 / variance :  0.9657230506657004
Residual sum of squares: 0.10
최고 평균 정확도: 0.9657
최적 파라미터: {'copy_X': True, 'fit_intercept': False, 'normalize': True}


In [79]:
grid.best_estimator_

LinearRegression(fit_intercept=False, normalize=True)

In [27]:
# from sklearn.metrics import accuracy_score
# best_clf = grid.best_estimator_
# pred = best_clf.predict(X_test)

In [ ]:
# acc = accuracy_score(y_test, pred)
# print(f'최적 파라미터 정확도: {acc:.4f}')
#에러; continuous is not supported

## sepal length의 weight, bias

In [89]:
X = df.iloc[:,[1,2,3,4]].values
y = df.iloc[:,0].values  #sepal length
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=2021
)
lr = LinearRegression()
lr.fit(X_train, y_train)
pred = lr.predict(X_test)
mse = mean_squared_error(y_test, pred)
r2 = r2_score(y_test, pred)
print(f'MSE: {mse:.2f}, RMSE: {np.sqrt(mse):.2f}')
print(f'R_squared: {r2:.4f}')

MSE: 0.07, RMSE: 0.27
R_squared: 0.8808


In [90]:
lr.coef_

array([ 0.60733316,  0.77406606, -0.49825608, -0.21265098])

In [91]:
lr.intercept_

1.9046012133683528

#### Sepal Length에 대한 회귀식
 - 0.6073\*sepal width + 0.7740\*petal length - 0.4982\*petal width - 0.2126\*class + 1.904

#### Sepal Length 교차검증

In [93]:
# 5 폴드 세트로 MSE를 구함
neg_mean_scores = cross_val_score(lr, X, y,
                                scoring='neg_mean_squared_error', cv=5)
rmse_scores = np.sqrt(-neg_mean_scores)
avg_rmse = np.average(rmse_scores)

print('개별 MSE: ',np.round(neg_mean_scores,2))
print('개별 RMSE: ',np.round(rmse_scores,2))
print(f'평균 RMSE: {avg_rmse:.4f}')

개별 MSE:  [-0.06 -0.08 -0.14 -0.1  -0.14]
개별 RMSE:  [0.25 0.29 0.37 0.31 0.38]
평균 RMSE: 0.3199


## sepal width의 weight, bias

In [94]:
X = df.iloc[:,[0,2,3,4]].values
y = df.iloc[:,1].values #sepal width
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=2021
)
lr = LinearRegression()
lr.fit(X_train, y_train)
pred = lr.predict(X_test)
mse = mean_squared_error(y_test, pred)
r2 = r2_score(y_test, pred)
print(f'MSE: {mse:.2f}, RMSE: {np.sqrt(mse):.2f}')
print(f'R_squared: {r2:.4f}')

MSE: 0.05, RMSE: 0.22
R_squared: 0.5667


In [95]:
lr.coef_

array([ 0.61339502, -0.59499152,  0.69803975, -0.12580651])

In [96]:
lr.intercept_

0.9976767818176961

#### Sepal Width에 대한 회귀식
- 0.6133\*sepal length - 0.5949\*petal length + 0.6980\*petal width - 0.1258\*class + 0.9976

#### Sepal Width 교차검증

In [97]:
# 5 폴드 세트로 MSE를 구함
neg_mean_scores = cross_val_score(lr, X, y,
                                scoring='neg_mean_squared_error', cv=5)
rmse_scores = np.sqrt(-neg_mean_scores)
avg_rmse = np.average(rmse_scores)

print('개별 MSE: ',np.round(neg_mean_scores,2))
print('개별 RMSE: ',np.round(rmse_scores,2))
print(f'평균 RMSE: {avg_rmse:.4f}')

개별 MSE:  [-0.08 -0.11 -0.16 -0.08 -0.11]
개별 RMSE:  [0.28 0.33 0.41 0.29 0.33]
평균 RMSE: 0.3289


## petal width의 weight, bias

In [85]:
X = df.iloc[:,[0,1,2,4]].values
y = df.iloc[:,3].values  #petal width
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=2021
)
lr = LinearRegression()
lr.fit(X_train, y_train)
pred = lr.predict(X_test)
mse = mean_squared_error(y_test, pred)
r2 = r2_score(y_test, pred)
print(f'MSE: {mse:.2f}, RMSE: {np.sqrt(mse):.2f}')
print(f'R_squared: {r2:.4f}')

MSE: 0.04, RMSE: 0.19
R_squared: 0.9226


In [86]:
lr.coef_

array([-0.12876754,  0.17861614,  0.32741898,  0.37562097])

In [87]:
lr.intercept_

-0.19445016041642238

#### Petal Width에 대한 회귀식
- -0.1287\*sepal length + 0.1786\*sepal width + 0.3274\*petal length + 0.3756\*class - 0.1944

#### Petal Width 교차검증

In [88]:
# 5 폴드 세트로 MSE를 구함
neg_mean_scores = cross_val_score(lr, X, y,
                                scoring='neg_mean_squared_error', cv=5)
rmse_scores = np.sqrt(-neg_mean_scores)
avg_rmse = np.average(rmse_scores)

print('개별 MSE: ',np.round(neg_mean_scores,2))
print('개별 RMSE: ',np.round(rmse_scores,2))
print(f'평균 RMSE: {avg_rmse:.4f}')

개별 MSE:  [-0.01 -0.02 -0.02 -0.04 -0.07]
개별 RMSE:  [0.09 0.14 0.15 0.19 0.26]
평균 RMSE: 0.1667
